File: scripts/dmzMBRAPluginNALinkObject.lua
local LinkHandle = dmz.handle.new ("Node_Link")

function update_object_position (self, ObjectHandle)
   local list = self.list[ObjectHandle]
   if list then
      for _, link in pairs (list) do
         local superPos = dmz.object.position (link.super)
         local subPos = dmz.object.position (link.sub)
         if superPos and subPos then
            local attrPos = superPos + ((subPos - superPos) * .5)
            dmz.object.position (link.attr, nil, attrPos)
         end
      end
   end
end

function destroy_object (self, handle)
   local list = self.list[handle]
   if list then
      for handle, link in pairs (list) do
         if self.list[link.super] then self.list[link.super][link] = nil end
         if self.list[link.sub] then self.list[link.sub][link] = nil end
      end
   end
end

function update_link_object (self, LinkHandle, AttrHandle, Super, Sub, AttrObj, PrevObj)
   if AttrObj then
      local link = { super = Super, sub = Sub, attr = AttrObj }
      if not self.list[Super] then self.list[Super] = {} end
      self.list[Super][LinkHandle] = link
      if not self.list[Sub] then self.list[Sub] = {} end
      self.list[Sub][LinkHandle] = link
      update_object_position (self, Super)
   elseif
         self.list[Super] and
         self.list[Super][LinkHandle] and
         self.list[Super][LinkHandle].attr == PrevObj then
      self.list[Super][LinkHandle] = nil
      self.list[Sub][LinkHandle] = nil
      dmz.object.destroy (PrevObj)
   end
end

function new (config, name)

   local self = {
      log = dmz.log.new ("lua." .. name),
      objObs = dmz.object_observer.new (),
      list = {},
   }

   self.log:info ("Creating plugin: " .. name)

   local cb = {
      update_object_position = update_object_position,
      destroy_object = destroy_object,
   }

   self.objObs:register (nil, cb, self)

   cb = { update_link_object = update_link_object, }
   self.objObs:register (LinkHandle, cb, self)

   return self
end

File: scripts/dmzMBRAPluginNALinkObject.js
var dmz =
      { object: require("dmz/components/object")
      , defs: require("dmz/runtime/definitions")
      }
   , LinkHandle = dmz.defs.createNamedHandle("Node_Link")
   , objectList = []
   ;

// Keep track of links, link labels, and the objects that are represented by link labels
// Whenever a node is moved, adjust all of the tags attached to its links to maintain a
// correct position
// Whenever a node is destroyed, remove all of the extraneous links, link objects, and
// link tags

// NOTE:
// Remember, JS does not allow operator overloading. Vectors cannot be added, subtracted,
// multiplied, etc, with mathematical operators -- these operations MUST be done via
// vector class methods, or else JS will try to convert the vectors to strings and
// perform the string-type versions of those operations, generally resulting in a NaN.
// The C++ bindings won't throw an error if a bad string is used.

var update_object_position = function (objectHandle) {
   var list = objectList[objectHandle]
     , attrPos
     , superPos
     , subPos
     , link
     ;
   if (list) {
      Object.keys(list).forEach(function (index) {
         link = list[index];
         superPos = dmz.object.position(link.superLink);
         subPos = dmz.object.position(link.sub);
         if (superPos && subPos) {
            attrPos = superPos.add((subPos.subtract(superPos)).multiply(0.5));
            dmz.object.position(link.attr, null, attrPos);
         }
      });
   }
};

dmz.object.position.observe(self, update_object_position);

dmz.object.destroy.observe(self, function (handle) {
   var list = objectList[handle]
     , link
     ;
   if (list) {
      Object.keys(list).forEach(function (index) {
         link = list[index];
         if (objectList[link.superLink]) {
            delete objectList[link.superLink][link.linkHandle];
         }
         if (objectList[link.sub]) {
            delete objectList[link.sub][link.linkHandle];
         }
      });
      delete objectList[handle];
   }
});

dmz.object.linkAttributeObject.observe(self, LinkHandle,
function (linkHandle, AttrHandle, Super, Sub, AttrObj, PrevObj) {
   var link
     ;
   if (AttrObj) {
      link = { superLink: Super, sub: Sub, attr: AttrObj, linkHandle: linkHandle };
      if (!objectList[Super]) {
         objectList[Super] = [];
      }
      objectList[Super][linkHandle] = link;
      if (!objectList[Sub]) {
         objectList[Sub] = [];
      }
      objectList[Sub][linkHandle] = link;

      update_object_position(Super);
   }
   else if (objectList[Super] && objectList[Super][linkHandle] &&
                 objectList[Super][linkHandle].attr == PrevObj) {
      delete objectList[Super][linkHandle];
      delete objectList[Sub][linkHandle];
   }
   if (PrevObj && dmz.object.isObject(PrevObj)) {
      dmz.object.destroy(PrevObj);
   }
});
